src/tqec/compile/substitute.py

```diff
@@ -0,0 +1,218 @@
+    # Substitute the final layer of the bottom block with the bulk layer
+    bottom_block = block1.with_updated_layer(
+        plaquettes_to_update=deepcopy(block1.layers[1].collection),
+        layer_to_update=block1.num_layers - 1,
+    )
+    # Substitute the first layer of the top block with the bulk layer
+    top_block = block2.with_updated_layer(
+        plaquettes_to_update=deepcopy(block2.layers[1].collection),
+        layer_to_update=0,
+    )
+    # If the pipe has a hadamard, apply the hadamard transformation
+    # The hadamard transformation is applied to the end of the bottom block
+    if key.pipe_type.has_hadamard:
+        for plaquette in bottom_block.layers[-1].collection.values():
+            _inplace_append_timelike_hadamard_transition(plaquette)
+    return bottom_block, top_block
+
+
+def _inplace_append_timelike_hadamard_transition(
+    plaquette: Plaquette,
+) -> None:
+    """Append a moment of transversal hadamard gates to the data qubits in the
+    plaquette."""
+    max_schedule_index = max(plaquette.circuit.schedule)
+    plaquette.circuit.add_to_schedule_index(
+        schedule=max_schedule_index + 1,
+        moment=cirq.Moment(
+            cirq.H(dq.to_grid_qubit()).with_tags(Plaquette._MERGEABLE_TAG)
+            for dq in plaquette.qubits.data_qubits
+        ),
+    )
+
+
+def _substitute_in_space_with_usual_cubes(
+    key: SubstitutionKey,
+    block1: CompiledBlock,
+    block2: CompiledBlock,
+) -> tuple[CompiledBlock, CompiledBlock]:
+    pipe_type = key.pipe_type
+    assert (
+        not key.spec1.is_spatial_junction and not key.spec2.is_spatial_junction
+    ), "Both cubes must not be spatial junction."
+    assert pipe_type.direction != Direction3D.Z, "Pipe direction must be X or Y."
+    # TODO: If the pipe has a hadamard, apply the spatial hadamard transformation
+    if pipe_type.has_hadamard:
+        raise TQECException("Hadamard on spatial pipes is not supported yet.")
+    substitute_side1 = (
+        PlaquetteSide.RIGHT
+        if pipe_type.direction == Direction3D.X
+        else PlaquetteSide.DOWN
+    )
+    substitute_side2 = substitute_side1.opposite()
+    # Get reset/mesurement basis
+    basis = pipe_type.value[2].upper()
+    block1 = _substitute_on_the_border(block1, substitute_side1, basis)
+    block2 = _substitute_on_the_border(block2, substitute_side2, basis)
+    return block1, block2
+
+
+def _substitute_on_the_border(
+    block: CompiledBlock,
+    plaquette_side: PlaquetteSide,
+    basis: str,
+) -> CompiledBlock:
+    if plaquette_side == PlaquetteSide.LEFT:
+        substitution = {1: 6, 8: 9, 7: 10}
+    elif plaquette_side == PlaquetteSide.UP:
+        substitution = {2: 12, 5: 10, 6: 9}
+    elif plaquette_side == PlaquetteSide.DOWN:
+        substitution = {3: 7, 14: 10, 13: 9}
+    else:  # plaquette_side == PlaquetteSide.RIGHT:
+        substitution = {4: 13, 11: 9, 12: 10}
```

Comment:
For the `QubitTemplate`, there is the systematic way to do this substitution. And the principle I used here is to copy the next-nearest neighbor plaquette in the opposite direction of  `plaquette_side` onto the border. This is because we just need to extend the bulk of the code along the direction of merge to make the merge operation valid. For example, when `plaquette_side == PlaquetteSide.LEFT`:

![图片](https://github.com/user-attachments/assets/82f15c5f-36f6-4599-8b95-855cc709b159)

I haven't look into the junctions yet and am not sure whether there is a simple systematic way too.

Reply:
Got it. It is not a generic way of doing (does not take into account that we will have to repeat specific plaquettes in the case of a spatial junction), but I think it is good enough for the CNOT.

Reply:
Yes, that's why I restrict its usage to `_substitute_in_space_with_usual_cubes`, which is only used when the two connected cubes are all "usual"(vs. spatial junction) .